### PR TITLE
Add PMD for static analysis

### PR DESCRIPTION
**Pull request by sladyn98**

### CODE REVIEW

**Comment by sladyn98** [comment]
This build fails because PMD throws errors, should I fix the errors in this PR itself ?

File: config/pmd.xml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+<ruleset name="Basic"
+    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
+  <description>
+The Basic ruleset contains a collection of good practices which should be followed.
+  </description>
+
+    <!-- Rules, that have been moved into a category -->
+    <rule ref="category/java/errorprone.xml/AvoidBranchingStatementAsLastInLoop" deprecated="true" />
```
Comment (martinda) [inline config/pmd.xml:12]: This syntax adds warnings to `mvn verify`. Please use the following one instead:
```xml
<rule ref="category/java/errorprone.xml">
      <exclude name="AvoidBranchingStatementAsLastInLoop" deprecated="true" />
      <!-- and so on -->
</rule>
```

**Comment by sladyn98** [comment]
@martinda These were thrown by findbugs so I fixed them else it was failing? Should I revert the commit and just let the pmd.xml remain ?

File: config/pmd.xml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    <rule ref="category/java/codestyle.xml/ExtendsObject" deprecated="true" />
+    <rule ref="category/java/codestyle.xml/ForLoopShouldBeWhileLoop" deprecated="true" />
+
+    <rule ref="category/java/performance.xml/BigIntegerInstantiation" deprecated="true" />
+    <rule ref="category/java/performance.xml/BooleanInstantiation" deprecated="true" />
+
+    <rule ref="category/java/design.xml/CollapsibleIfStatements" deprecated="true" />
+    <rule ref="category/java/design.xml/SimplifiedTernary" deprecated="true" />
+
+</ruleset>
```
Comment (martinda) [inline config/pmd.xml:42]: Actually, the rule set needs to be written like this:
```
    <rule ref="category/java/bestpractices.xml" />
    <rule ref="category/java/codestyle.xml" />
    <rule ref="category/java/design.xml" />
    <rule ref="category/java/documentation.xml" />
    <rule ref="category/java/errorprone.xml" />
    <rule ref="category/java/multithreading.xml" />
    <rule ref="category/java/performance.xml" />
    <rule ref="category/java/security.xml" />
```
Then run and fix as many as possible. Some cannot be fixed or are not that critical (like upper case naming conventions), and those can be excluded.

File: pom.xml [hunk 3]
```diff
@@ -95,6 +95,41 @@
...
+						<goals>
+							<goal>check</goal>
+							<goal>cpd-check</goal>
+						</goals>
+					</execution>
+				</executions>
+			</plugin>
+				<plugin>
+				<groupId>org.codehaus.mojo</groupId>
+				<artifactId>findbugs-maven-plugin</artifactId>
```
Comment (martinda) [inline pom.xml:118]: Findbugs should be its own PR. It makes the review easier to do for reviewers.

**Comment by martinda** [comment]
> @martinda These were thrown by findbugs so I fixed them else it was failing? Should I revert the commit and just let the pmd.xml remain ?

To make things easy and clear, please present two pull-requests. One for PMD, one for findbugs. Please do not lose my comments on PMD. So yes it is probably best to move findbugs to another PR and do just PMD in this one.

File: config/pmd.xml (hunk 2 above)
Comment (kwhetstone) [reply config/pmd.xml:42]: I'm not sure that you need to add "deprecated" to the end of each of these calls either.

File: pom.xml [hunk 4]
```diff
@@ -255,5 +290,4 @@
 		  <url>https://repo.jenkins-ci.org/public/</url>
 		</pluginRepository>
 	</pluginRepositories>
-
 </project>
```
Comment (kwhetstone) [inline pom.xml:293]: Add some new lines at the end your files.  It makes it look cleaner.

File: src/main/java/com/org/jenkins/custom/jenkins/distribution/service/generators/WarGenerator.java [hunk 5]
```diff
@@ -18,7 +19,7 @@ public static File generateWAR(String versionName, String configuration) throws
         final Config cfg;
         Path tempDirWithPrefix = Files.createTempDirectory(TEMP_PREFIX);
             File packagerConfigFile = File.createTempFile("packager-config", ".yml");
-            byte[] buf = configuration.getBytes();
+            byte[] buf = configuration.getBytes(StandardCharsets.UTF_8);
```
Comment (kwhetstone) [inline src/main/java/com/org/jenkins/custom/jenkins/distribution/service/generators/WarGenerator.java:22]: Was this found by the scans?

File: config/pmd.xml (hunk 1 above)
Comment (sladyn98) [reply config/pmd.xml:12]: Sure :+1:

**Comment by sladyn98** [comment]
@kwhetstone @martinda Yeah the findbugs is causing the build to fail hence I fixed them here.

File: config/pmd.xml (hunk 2 above)
Comment (sladyn98) [reply config/pmd.xml:42]: Yeah

File: src/main/java/com/org/jenkins/custom/jenkins/distribution/service/generators/WarGenerator.java (hunk 5 above)
Comment (sladyn98) [reply src/main/java/com/org/jenkins/custom/jenkins/distribution/service/generators/WarGenerator.java:22]: Yeah  I will put them back to normal

File: config/pmd.xml (hunk 2 above)
Comment (martinda) [reply config/pmd.xml:35]: Writing the rules like that causes lots of warnings:
```
[WARNING] Unable to locate Source XRef to link to - DISABLED
[WARNING] Unable to locate Source XRef to link to - DISABLED
[WARNING] The rule AvoidDecimalLiteralsInBigDecimalConstructor is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule AvoidMultipleUnaryOperators is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule AvoidUsingOctalValues is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule BrokenNullCheck is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule CheckSkipResult is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule ClassCastExceptionWithToArray is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule DontUseFloatTypeForLoopIndices is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule JumbledIncrementer is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule MisplacedNullCheck is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule OverrideBothEqualsAndHashcode is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule ReturnFromFinallyBlock is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule UnconditionalIfStatement is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule AvoidDecimalLiteralsInBigDecimalConstructor is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule AvoidMultipleUnaryOperators is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule AvoidUsingOctalValues is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule BrokenNullCheck is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule CheckSkipResult is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule ClassCastExceptionWithToArray is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule DontUseFloatTypeForLoopIndices is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule JumbledIncrementer is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule MisplacedNullCheck is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule OverrideBothEqualsAndHashcode is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule ReturnFromFinallyBlock is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule UnconditionalIfStatement is referenced multiple times in "Basic". Only the last rule configuration is used.
```

Please use the syntax I showed, it will eliminate all warnings. And put one rule per line, meaning that if you have an exception, write it on another line:
```
<rule ref="category/java/errorprone.xml">
     <exclude name="AvoidBranchingStatementAsLastInLoop" />
</rule>
```

File: src/main/java/com/org/jenkins/custom/jenkins/distribution/service/services/UpdateCenterService.java [hunk 6]
```diff
@@ -37,14 +36,15 @@ public JSONObject downloadUpdateCenterJSON() throws Exception {
...
             LOGGER.info("Creating a new file" + updateCenterFile.getPath());
-            HttpGet get = new HttpGet(UPDATE_CENTER_JSON_URL);
             LOGGER.info("Executing Request at " + UPDATE_CENTER_JSON_URL);
-            CloseableHttpClient httpClient = HttpClients.createDefault();
-            CloseableHttpResponse response = httpClient.execute(get);
-            responseString = EntityUtils.toString(response.getEntity());
-            byte[] buf = responseString.getBytes(StandardCharsets.UTF_8);
+            try (CloseableHttpClient httpClient = HttpClients.createDefault();
+                 CloseableHttpResponse response = httpClient.execute(new HttpGet(UPDATE_CENTER_JSON_URL))) {
+                responseString = EntityUtils.toString(response.getEntity());}
```
Comment (martinda) [inline src/main/java/com/org/jenkins/custom/jenkins/distribution/service/services/UpdateCenterService.java:42]: Please fix the indentation here, it will clarify the intent of the try-with-resources statement.

Comment (sladyn98) [reply src/main/java/com/org/jenkins/custom/jenkins/distribution/service/services/UpdateCenterService.java:42]: Thanks Martin working on them

**Comment by kwhetstone** [comment]
You have 58 PMD violations. For more details see: C:\Jenkins\workspace\stom-distribution-service_PR-102\target\pmd.xml ?

**Comment by sladyn98** [comment]
@kwhetstone yeah working on fixing as much as I can slowly , cuz some of them are a real pain
